Improvements to Label's layout options

Geometror opened the pull request:
This PR aims to improve the layout options for the Label control. 
Implements godotengine/godot-proposals#2916 and godotengine/godot-proposals#2931
(This needed to be done at once since the additions are interdependent)
If these additions/changes get approved, it makes sense to add them (partly/fully) to TextLine and TextParagraph in a subsequent PR (or this?).
![grafik](https://user-images.githubusercontent.com/50084500/124296612-73b95f00-db5a-11eb-9fcf-6bc0525e5d37.png)


- Added options to trim the text in case it overruns (exceeds the node's bounding box)
    - Trim characters:
![Animation1](https://user-images.githubusercontent.com/50084500/124299265-85e8cc80-db5d-11eb-99c2-e9ba20f454dd.gif)

    - Trim words:
![Animation2](https://user-images.githubusercontent.com/50084500/124299290-8c774400-db5d-11eb-82dc-33fb395a5ff1.gif)

    - Ellipsis trimming:
![Animation3](https://user-images.githubusercontent.com/50084500/124299292-8ed99e00-db5d-11eb-9d7e-539092553bd9.gif)

    - Ellipsis word trimming:
![Animation4](https://user-images.githubusercontent.com/50084500/124299302-91d48e80-db5d-11eb-9054-ff73fefc39a1.gif)


- Added more autowrap modes
    - Arbitrary:
![Animation5](https://user-images.githubusercontent.com/50084500/124299956-4ec6eb00-db5e-11eb-864e-df65fc2ed372.gif)
    - Word (Smart):
![Animation7](https://user-images.githubusercontent.com/50084500/124300048-68683280-db5e-11eb-95c3-d97b8449f7f5.gif)
- Slight change to line breaking logic, which ignores single trailing spaces

New properties:
![grafik](https://user-images.githubusercontent.com/50084500/124301350-fd1f6000-db5f-11eb-8407-9abb87f15089.png)
![grafik](https://user-images.githubusercontent.com/50084500/124301369-06103180-db60-11eb-8a3e-c0eb7339cd2b.png)

The only thing missing is proper RTL text support (the ellipsis string should be placed on the left then), but that will be added once I figured out how to do that properly. Also, I am not sure if I implemented it correctly or at the right place in the codebase but after digging through the code it seemed like the only option without changing huge parts of the code.

Any testing/feedback would be appreciated :)

bruvzg commented on modules/text_server_adv/text_server_adv.cpp:
```diff
@@ -1658,6 +1658,123 @@ float TextServerAdvanced::shaped_text_tab_align(RID p_shaped, const Vector<float
+	int last_valid_cut = 0;
+	bool found = false;
+
+	for (int i = sd_size - 1; i >= 0; --i) {
```
This should run from the other end (and add ellipsis on the left side) if detected text direction (`sd->para_direction`) is RTL.

<img width="380" alt="Screenshot 2021-07-02 at 19 23 47" src="https://user-images.githubusercontent.com/7645683/124303672-6062bf80-db6b-11eb-9437-ec3e3634b4b0.png">

Geometror replied:
Implemented.

bruvzg commented on scene/gui/label.cpp:
```diff
@@ -370,13 +420,13 @@ Size2 Label::get_minimum_size() const {
 			min_size.width = 1;
 		}
-
+		auto res = min_size + min_style;
```
Using `auto` is usually against Godot code style.

Geometror replied:
Oops, that line was for debugging and I forgot to remove it. Fixed.

bruvzg commented on servers/text_server.h:
```diff
@@ -66,8 +66,15 @@ class TextServer : public Object {
+		BREAK_WORD_BOUND_ADAPTIVE = 1 << 5 | 1 << 7
+	};
+
+	enum TextOverrunFlag {
```
Missing `VARIANT_ENUM_CAST(TextServer::TextOverrunFlag);` at the end of file and `BIND_ENUM_CONSTANT(...)`s in the text server `_bind_methods()`.

Geometror replied:
Fixed.

bruvzg commented on modules/text_server_adv/text_server_adv.cpp:
```diff
@@ -1658,6 +1658,123 @@ float TextServerAdvanced::shaped_text_tab_align(RID p_shaped, const Vector<float
+				gl.index = whitespace_gl_idx;
+				gl.font_rid = last_gl_font_rid;
+				gl.font_size = last_gl_font_size;
+				gl.flags = GRAPHEME_IS_SPACE | GRAPHEME_IS_BREAK_SOFT;
```
The added glyphs probably should have `GRAPHEME_IS_VIRTUAL` bit set (to be non-selectable and skipped by caret movement), irrelevant for the label, but good for consistency.

Geometror replied:
Done.

Geometror commented:
Added RTL support and did some optimizations/refactoring
![grafik](https://user-images.githubusercontent.com/50084500/124366736-1e607900-dc52-11eb-89c5-5bd75b05801a.png)

Geometror committed:
Improvements to Label's layout options
- Added options to trim the text in case it overruns
- Added more autowrap modes
- Improved line breaking, which ignores trailing spaces

Geometror commented on modules/text_server_adv/text_server_adv.cpp:
```diff
@@ -1658,6 +1658,161 @@ float TextServerAdvanced::shaped_text_tab_align(RID p_shaped, const Vector<float
+			sd->glyphs.resize(trim_pos + added_glyphs);
+		} else {
+			for (int ridx = 0; ridx <= trim_pos - added_glyphs; ridx++) {
+				sd->glyphs.remove(0);
```
Done. I also had to fix another bug introduced by former optimizations in the code for handling RTL text.

akien-mga commented:
Thanks!
